File: sdks/node/src/transport.ts
// Copyright © 2026 Jalapeno Labs

import type { Error as ContractError } from './proto/arsox/error/v1/error_pb.js'
import type { DescMessage, MessageInitShape, MessageShape } from '@bufbuild/protobuf'

// Core
import { request as httpRequest } from 'node:http'
import { request as httpsRequest } from 'node:https'
import { create, fromBinary, toBinary } from '@bufbuild/protobuf'

// Misc
import { ErrorSchema } from './proto/arsox/error/v1/error_pb.js'
import { ArsoxError } from './error.js'
import { PROTOBUF_CONTENT_TYPE } from './constants.js'

/** The methods the satellite's HTTP surface uses. */
export type HttpMethod = 'GET' | 'POST' | 'DELETE'

/** What came back off the socket, before anything is made of it. */
export type RawResponse = {
  status: number
  body: Uint8Array
}

/**
 * Encodes a message for the wire.
 *
 * `create` then `toBinary` is the protobuf-es pair, and writing it out at every
 * call site means naming the schema twice for one message, which is exactly the
 * kind of duplication that eventually names two different schemas.
 */
export function encode<Desc extends DescMessage>(
  schema: Desc,
  init: MessageInitShape<Desc>
): Uint8Array {
  return toBinary(schema, create(schema, init))
}

/**
 * Turns a raw response into the message it carries, or into the failure it
 * describes.
 *
 * Separated from the socket work so the mapping from a status and some bytes to
 * an {@link ArsoxError} is testable without a satellite, a port, or a mock.
 */
export function decodeResponse<Desc extends DescMessage>(
  schema: Desc,
  response: RawResponse
): MessageShape<Desc> {
  if (response.status >= 200 && response.status < 300) {
    try {
      return fromBinary(schema, response.body)
    }
    catch (error) {
      throw ArsoxError.transport(
        `the satellite sent an undecodable response: ${String(error)}`
      )
    }
  }

  // Every failure on every transport is the same shape, so an empty or
  // undecodable error body means something other than a satellite answered.
  if (response.body.byteLength === 0) {
    throw ArsoxError.transport(`the satellite answered ${response.status} with no body`)
  }

  let contractError: ContractError
  try {
    contractError = fromBinary(ErrorSchema, response.body)
  }
  catch {
    throw ArsoxError.transport(`the satellite answered ${response.status}`)
  }

  throw ArsoxError.contract(contractError)
}

/**
 * One satellite's address and credential.
 *
 * Handles hold one of these and nothing else, which is what makes a handle
 * disposable and a thread durable.
 */
export class Connection {
  /** The satellite's base URL, without a trailing slash. */
  readonly baseUrl: string

  readonly secret: string

  constructor(url: string, secret: string) {
    this.baseUrl = url.replace(/\/+$/u, '')
    this.secret = secret
  }

  /** The same satellite, addressed as a WebSocket. */
  get socketUrl(): string {
    if (this.baseUrl.startsWith('https://')) {
      return `wss://${this.baseUrl.slice('https://'.length)}`
    }
    if (this.baseUrl.startsWith('http://')) {
      return `ws://${this.baseUrl.slice('http://'.length)}`
    }
    return this.baseUrl
  }

  /**
   * Sends one request and decodes what comes back.
   *
   * `body` is optional because several endpoints take none, and present on GET
   * for the listings, whose filters are a protobuf request message like every
   * other request in the contract.
   */
  async call<Desc extends DescMessage>(
    method: HttpMethod,
    path: string,
    schema: Desc,
    body?: Uint8Array
  ): Promise<MessageShape<Desc>> {
    return decodeResponse(schema, await this.transmit(method, path, body))
  }

  /**
   * Puts one request on the wire.
   *
   * Built on `node:http` rather than `fetch`. The contract carries filters for
   * the listing endpoints in a protobuf request body on GET, and the WHATWG
   * fetch specification forbids a body on GET outright, so `fetch` throws before
   * a byte leaves the process. This is not a preference for the older API.
   */
  private transmit(
    method: HttpMethod,
    path: string,
    body: Uint8Array | undefined
  ): Promise<RawResponse> {
    const url = new URL(this.baseUrl + path)
    const send = url.protocol === 'https:'
      ? httpsRequest
      : httpRequest

    const headers: Record<string, string> = {
      Authorization: `Bearer ${this.secret}`,
      Accept: PROTOBUF_CONTENT_TYPE
    }
    if (body) {
      headers['Content-Type'] = PROTOBUF_CONTENT_TYPE
      headers['Content-Length'] = String(body.byteLength)
    }

    return new Promise((resolve, reject) => {
      const outgoing = send(url, { method, headers }, (incoming) => {
        const chunks: Buffer[] = []

        incoming.on('data', (chunk: Buffer) => chunks.push(chunk))
        incoming.on('error', (error: Error) => reject(ArsoxError.transport(error.message)))
        incoming.on('end', () => resolve({
          status: incoming.statusCode ?? 0,
          body: Buffer.concat(chunks)
        }))
      })

      outgoing.on('error', (error: Error) => reject(ArsoxError.transport(error.message)))

      if (body) {
        outgoing.write(body)
      }
      outgoing.end()
    })
  }
}

File: sdks/python/src/arsox_sdk/transport.py
# Copyright © 2026 Jalapeno Labs

"""One satellite's address, credential, and connection pool."""

from __future__ import annotations

import logging
from typing import Literal, TypeVar

import aiohttp
from google.protobuf.message import DecodeError, Message

from arsox_sdk.constants import PROTOBUF_CONTENT_TYPE
from arsox_sdk.errors import ArsoxError
from arsox_sdk.proto.arsox.error.v1 import error_pb2

logger = logging.getLogger(__name__)

# The methods the satellite's HTTP surface uses.
HttpMethod = Literal["GET", "POST", "DELETE"]

# The thread stream's socket, opened with text decoding off. Frames are binary
# protobuf, always, so a text frame is never something this SDK reads.
#
# Parameterized on `bool` rather than on `Literal[False]` because that is what
# aiohttp's own overloads resolve a keyword `decode_text=False` to.
WebSocket = aiohttp.ClientWebSocketResponse[bool]

_Response = TypeVar("_Response", bound=Message)


def decode_response(message_type: type[_Response], status: int, body: bytes) -> _Response:
    """Turn a raw response into the message it carries, or the failure it describes.

    Separated from the socket work so the mapping from a status and some bytes to
    an `ArsoxError` is testable without a satellite, a port, or a mock.

    Args:
        message_type: the contract message the caller expects on success.
        status: the HTTP status the satellite answered with.
        body: the response body, protobuf either way.

    Raises:
        ArsoxError: on any status outside 2xx, and on a success body that will
            not decode.
    """
    if 200 <= status < 300:
        message = message_type()
        try:
            message.ParseFromString(body)
        except DecodeError as error:
            raise ArsoxError.transport(
                f"the satellite sent an undecodable response: {error}"
            ) from error

        return message

    # Every failure on every transport is the same shape, so an empty or
    # undecodable error body means something other than a satellite answered.
    if not body:
        raise ArsoxError.transport(f"the satellite answered {status} with no body")

    contract_error = error_pb2.Error()
    try:
        contract_error.ParseFromString(body)
    except DecodeError as error:
        raise ArsoxError.transport(f"the satellite answered {status}") from error

    raise ArsoxError.contract(contract_error)


class Connection:
    """One satellite's address, credential, and connection pool.

    Handles hold one of these and nothing else, which is what makes a handle
    disposable and a thread durable.
    """

    def __init__(self, url: str, secret: str, session: aiohttp.ClientSession) -> None:
        """Point a connection at a satellite."""
        self.base_url = url.rstrip("/")
        self.secret = secret
        self.session = session

    @property
    def socket_url(self) -> str:
        """The same satellite, addressed as a WebSocket."""
        if self.base_url.startswith("https://"):
            return f"wss://{self.base_url[len('https://'):]}"
        if self.base_url.startswith("http://"):
            return f"ws://{self.base_url[len('http://'):]}"
        return self.base_url

    async def call(
        self,
        method: HttpMethod,
        path: str,
        message_type: type[_Response],
        body: bytes | None = None,
    ) -> _Response:
        """Send one request and decode what comes back.

        `body` is optional because several endpoints take none, and present on
        GET for the listings, whose filters are a protobuf request message like
        every other request in the contract. aiohttp puts a body on a GET; the
        WHATWG fetch specification forbids one outright, which is the constraint
        that shaped the Node client and would have shaped this one.

        Raises:
            ArsoxError: when the satellite is unreachable, rejects the request,
                or answers something undecodable.
        """
        headers = {
            "Authorization": f"Bearer {self.secret}",
            "Accept": PROTOBUF_CONTENT_TYPE,
        }
        if body is not None:
            headers["Content-Type"] = PROTOBUF_CONTENT_TYPE

        try:
            async with self.session.request(
                method, self.base_url + path, headers=headers, data=body
            ) as response:
                return decode_response(message_type, response.status, await response.read())
        except aiohttp.ClientError as error:
            raise ArsoxError.transport(str(error)) from error

    async def open_socket(self, path: str) -> WebSocket:
        """Open one of the satellite's WebSockets.

        Awaited rather than lazy, so a refused handshake is reported where the
        caller asked for the stream rather than at the first event.

        Raises:
            ArsoxError: when the handshake is refused or the satellite is
                unreachable.
        """
        try:
            return await self.session.ws_connect(
                self.socket_url + path,
                headers={"Authorization": f"Bearer {self.secret}"},
                decode_text=False,
            )
        except aiohttp.ClientError as error:
            raise ArsoxError.transport(str(error)) from error

    async def close(self) -> None:
        """Release the connection pool."""
        await self.session.close()
